src/v3/plugins/git/graphView.js

```diff
@@ -0,0 +1,333 @@
+      const commit: GN.CommitAddress = (GN.fromRaw(rawAddress): any);
+      this._maybeCheckInvariants();
+      yield commit;
+    }
+    this._maybeCheckInvariants();
+  }
+
+  tree(commit: GN.CommitAddress): GN.TreeAddress {
+    const result: GN.TreeAddress = Array.from(
+      this._neighbors(commit, {
+        direction: Direction.OUT,
+        nodePrefix: GN._Prefix.tree,
+        edgePrefix: GE._Prefix.hasTree,
+      })
+    )[0];
+    this._maybeCheckInvariants();
+    return result;
+  }
+
+  parents(commit: GN.CommitAddress): Iterator<GN.CommitAddress> {
+    const result: Iterator<GN.CommitAddress> = this._neighbors(commit, {
+      direction: Direction.OUT,
+      nodePrefix: GN._Prefix.commit,
+      edgePrefix: GE._Prefix.hasParent,
+    });
+    this._maybeCheckInvariants();
+    return result;
+  }
+
+  entries(tree: GN.TreeAddress): Iterator<GN.TreeEntryAddress> {
+    const result: Iterator<GN.TreeEntryAddress> = this._neighbors(tree, {
+      direction: Direction.OUT,
+      nodePrefix: GN._Prefix.treeEntry,
+      edgePrefix: GE._Prefix.includes,
+    });
+    this._maybeCheckInvariants();
+    return result;
+  }
+
+  contents(entry: GN.TreeEntryAddress): Iterator<GN.TreeEntryContentsAddress> {
+    const result: Iterator<GN.TreeEntryContentsAddress> = this._neighbors(
+      entry,
+      {
+        direction: Direction.OUT,
+        nodePrefix: GN._Prefix.base, // multiple kinds
+        edgePrefix: GE._Prefix.hasContents,
+      }
+    );
+    this._maybeCheckInvariants();
+    return result;
+  }
+
+  evolvesTo(entry: GN.TreeEntryAddress): Iterator<GN.TreeEntryAddress> {
+    const result: Iterator<GN.TreeEntryAddress> = this._neighbors(entry, {
+      direction: Direction.OUT,
+      nodePrefix: GN._Prefix.treeEntry,
+      edgePrefix: GE._Prefix.becomes,
+    });
+    this._maybeCheckInvariants();
+    return result;
+  }
+
+  evolvesFrom(entry: GN.TreeEntryAddress): Iterator<GN.TreeEntryAddress> {
+    const result: Iterator<GN.TreeEntryAddress> = this._neighbors(entry, {
+      direction: Direction.IN,
+      nodePrefix: GN._Prefix.treeEntry,
+      edgePrefix: GE._Prefix.becomes,
+    });
+    this._maybeCheckInvariants();
+    return result;
+  }
+
+  _maybeCheckInvariants() {
+    if (process.env.NODE_ENV === "test") {
+      // TODO(perf): If this method becomes really slow, we can disable
+      // it on specific tests wherein we construct large graphs.
+      this.checkInvariants();
+    }
+  }
+
+  checkInvariants() {
+    // All Git nodes and edges must have valid Git addresses.
+    for (const node of this._graph.nodes({prefix: GN._Prefix.base})) {
+      GN.fromRaw((((node: NodeAddressT): any): GN.RawAddress));
+    }
+    for (const edge of this._graph.edges({
+      addressPrefix: GE._Prefix.base,
+      srcPrefix: NodeAddress.empty,
+      dstPrefix: NodeAddress.empty,
+    })) {
+      GE.fromRaw((((edge.address: EdgeAddressT): any): GE.RawAddress));
```

Comment:
nit: This gets checked again on line 195.

Reply:
It does indeed. I think I’ll remove this one: I kept it so that the node
and edge checks were adjacent, but I don’t really see any reason to do
so, and it needlessly wastes cycles.